compiler/src/iree/compiler/Dialect/VM/IR/test/comparison_folding.mlir

```diff
@@ -35,6 +35,60 @@ vm.module @cmp_eq_i32_folds {
 
 // -----
 
+
+// CHECK-LABEL: @cmp_eq_f32_near_folds
+vm.module @cmp_eq_f32_near_folds {
+  // CHECK-LABEL: @eq_near_f32
+  vm.func @eq_near_f32(%arg0 : f32, %arg1 : f32) -> i32 {
+    // CHECK-NEXT:   %zero = vm.const.f32.zero
+    // CHECK-NEXT:   %c1 = vm.const.i32 1
+    // CHECK-NEXT:   [[THRESHOLD:%.+]] = vm.const.i32
+    // CHECK-NEXT:   %0 = vm.cmp.lt.f32.o %arg0, %zero : f32
+    // CHECK-NEXT:   %1 = vm.xor.i32 %0, %c1 : i32
+    // CHECK-NEXT:   %2 = vm.cmp.lt.f32.o %arg1, %zero : f32
+    // CHECK-NEXT:   %3 = vm.xor.i32 %2, %c1 : i32
+    // CHECK-NEXT:   %ne = vm.cmp.ne.i32 %1, %3 : i32
+    // CHECK-NEXT:   vm.cond_br %ne, ^bb1, ^bb2
+    // CHECK-NEXT: ^bb1:  // pred: ^bb0
+    // CHECK-NEXT:   %4 = vm.cmp.eq.f32.o %arg0, %arg1 : f32
+    // CHECK-NEXT:   vm.br ^bb3(%4 : i32)
+    // CHECK-NEXT: ^bb2:  // pred: ^bb0
+    // CHECK-NEXT:   %5 = vm.bitcast.f32.i32 %arg0 : f32 -> i32
+    // CHECK-NEXT:   %6 = vm.bitcast.f32.i32 %arg1 : f32 -> i32
+    // CHECK-NEXT:   %7 = vm.sub.i32 %5, %6 : i32
+    // CHECK-NEXT:   %8 = vm.abs.i32 %7 : i32
+    // CHECK-NEXT:   %slt = vm.cmp.lt.i32.s %8, [[THRESHOLD]] : i32
+    // CHECK-NEXT:   vm.br ^bb3(%slt : i32)
+    // CHECK-NEXT: ^bb3(%9: i32):  // 2 preds: ^bb1, ^bb2
+    // CHECK-NEXT:   vm.return %9 : i32
+    %cmp = vm.cmp.eq.f32.near %arg0, %arg1 : f32
+    vm.return %cmp : i32
+  }
+
+  // TODO(scotttodd): figure out why this goes through the folder correctly
+  //                  but canonicalizes anyways (check return types?)
+  // CHECK-LABEL: @const_eq_near_f32
+  vm.func @const_eq_near_f32() -> i32 {
+    // CHECK: %zero = vm.const.i32.zero
+    // CHECK-NEXT: vm.return %zero
+    %a = vm.const.f32 1.9999999
+    %b = vm.const.f32 2.0
+    %cmp = vm.cmp.eq.f32.near %a, %b : f32
+    vm.return %cmp : i32
```

Comment:
I think there's a preexisting bug somewhere in the VM op folders that I'm hitting while trying to test this new code. I can reproduce similar behavior in `CmpGTF32OOp::fold`. When I dump the `OpFoldResult`, I see `1.000000e+00 : f32`, leading to the pattern failing to match. If I similarly dump the result in `CmpEQF32OOp::fold`, I see `1 : i32`, which matches as expected.

Reply:
A bit more specifically, I suspect that this code: https://github.com/iree-org/iree/blob/2c5344a4bdc1d5cba40fd8148b779e3e1ad47b0c/compiler/src/iree/compiler/Dialect/VM/IR/VMOpFolders.cpp#L580-L587 is not set up to handle ops with different argument and result types, such as https://github.com/iree-org/iree/blob/2c5344a4bdc1d5cba40fd8148b779e3e1ad47b0c/compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td#L2735-L2741

(It's been a while since I've touched this compiler code, so I'm a bit fuzzy on a few concepts)

Reply:
Yep, that's the issue! I hacked `constFoldBinaryOp` to return `IntegerAttr::get(op.result().getType(), ...)` and the folders work as expected. I'll see if I can find a clean way to work that into the existing templates/args, or I can add a separate function for folding ops that have differing argument and result types.